src/libcore/char/methods.rs

```diff
@@ -455,9 +455,13 @@ impl char {
                 *dst.get_unchecked_mut(3) = (code & 0x3F) as u8 | TAG_CONT;
                 4
             } else {
+                // We know that all chars fit inside 4 bytes
+                if dst.len() <= 4 {
```

Comment:
Shouldn't this be `< 4`?

Reply:
or rather `>= 4`
example: if `self` requires more than one byte to be encoded and `dst.len` is 1, this code is reached and should not be considered unreachable.

Reply:
In the old code it wasn't reachable. After the slice changes that came after my comment I need to reevaluate.

Reply:
I'm re-verifying... I think @ranma42 might be right, but in any case, I think I found an LLVM bug -- https://rust.godbolt.org/z/ddVHt8 should optimize out, i.e., get an empty `bar` function and so on, since there's no way for the panic to be hit, but doesn't. That might make getting the optimization to fire here quite difficult.

I'm attempting now to figure out whether we can do something in Rust to convince LLVM.